sqe-http-api/Helpers/SystemEmailer.cs

```diff
@@ -2,56 +2,93 @@
 using System.Threading.Tasks;
 using MailKit.Security;
 using Microsoft.AspNetCore.Identity.UI.Services;
+using Microsoft.Extensions.Configuration;
 using MimeKit;
-using SmtpClient = MailKit.Net.Smtp.SmtpClient;
+using MailKit.Net.Smtp;
+using Microsoft.AspNetCore.Hosting;
+using SQE.SqeHttpApi.DataAccess.Helpers;
 
 namespace SQE.SqeHttpApi.Server.Helpers
 {
     public class DevEmailSender : IEmailSender
     {
+        private readonly IConfiguration _config;
+        private readonly IHostingEnvironment _env;
+
+        public DevEmailSender(IConfiguration config, IHostingEnvironment env)
+        {
+            _config = config;
+            _env = env;
+        }
+        
         /// <summary>
-        /// Sends an email.  This relies on the environment variables SQE_GMAIL_USERNAME and SQE_GMAIL_PASSWORD.
-        /// If you do not set these, the function will error.
-        /// TODO: Figure out what email server we might use in production.  Alter this to be more flexible regarding
-        /// email provider and account.
+        /// Sends an email.  This uses the settings in appsettings.json to send the email via an external SMTP server.
+        /// When this program is run in its docker container, the environment variables can be used to provide custom
+        /// settings on container start (automatically done via startup.sh).  The environment variables are:
+        /// MAILER_EMAIL_ADDRESS, MAILER_EMAIL_USERNAME, MAILER_EMAIL_PASSWORD, MAILER_EMAIL_SMTP_URL,
+        /// MAILER_EMAIL_SMTP_PORT, MAILER_EMAIL_SMTP_SECURITY (this should be a string corresponding to one of the
+        /// options in the SecureSocketOptions enum).
         /// </summary>
         /// <param name="email">Email address to send the message to.</param>
         /// <param name="subject"></param>
         /// <param name="htmlMessage"></param>
         /// <returns></returns>
         public async Task SendEmailAsync(string email, string subject, string htmlMessage)
         {
-            try  
+            // Don't bother trying to send an email unless we have all the necessary email settings
+            // TODO: Should we throw an error for this, or even kill the whole program?
+            if (string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailAddress"))
+                || string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailUsername"))
+                || string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailPassword"))
+                || string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailSmtpUrl"))
+                || string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailSmtpPort"))
+                || string.IsNullOrEmpty(_config.GetConnectionString("MailerEmailSmtpSecurity"))
+                )
+                return;
```

Comment:
Didn't we say we don't want such configuration errors to be swallowed silently?

Reply:
Yeah, that is what I was wondering about in the TODO question.  I guess, my feeling is that in production I think I want to exit the program if no credentials are available for the SMTP server. In development, I am happy to let the program run without any SMTP credentials at all (since I am often working on other things and don't care about the email system). So I guess, I should completely remove these checks from the emailer and instead put some logic near startup that checks them on startup when running in production mode (and gracefully exits with a clear error message when they are not).

Reply:
The server can run on development without the email, but you must return an error if someone tries to register a new user on a development machine with no email. They will not be able to do that, and they need to see it. Otherwise we can expect 5 or 6 lost hours figuring out where are all those activation emails.

Reply:
Ok.  I made the change now.  I put in checks in the `Startup` method in `Startup.cs`, but that feels wrong to me.  Where would you put such checks to verify that the settings in appsettings.json are valid?  I use `if (env.IsProduction())` to only run these tests in production.  Also what would you do to stop the program in the case that one of these checks finds an unset variable, I used `throw new SystemException` which crashes the program as it is an uncaught exception.

On the other hand. I could forget those checks on startup altogether and just let MailKit throw it's error; there is no guarantee after all that the settings in appsettings.json are correct just because they are there.  It is setup now in SystemEmailer.cs that in production it will catch any send errors and throw a nice message back to the user saying they probably input an invalid email (I still need to add a server-side log), otherwise in development it sends the whole error on.

Reply:
That's a good question. Actually, that's a few good questions. First - about the placement of the code, putting it in Startup.cs (or at least - calling it from Startup.cs) seems OK. 

The bigger question is which checks we're going to perform at startup, so that we know the server is not properly configured right away, instead of later on. There are a few things that need to be configured properly: the database, redis for SignalR, the email configuration and perhaps more in the future.

I think it's more important to add logging than to check these things at startup - because these thing can fail after startup, too (the mail server can die, for instance). I guess at startup we just need to check that the configuration is complete, and log a warning if it isn't on dev, or stop the server on prod (and log a critical error).

Reply:
Many thanks.  I will work on implementing that now.

Reply:
I now created a class `StartupChecks` which gets run at startup.  It always checks that that database can be accessed, and when running in production mode it tests that it can connect to the SMTP server.  In both cases it not only checks for the necessary settings in appsettings.json, but also makes a real connection to the external resource.  So that should help in killing the program immediately if the settings are invalid or the resource is unreachable.  Any runtime errors are caught in the application code now and will manifest in errors sent to the client, but I have not yet added server side logging for them.  I don't want to add logging to this PR, I think we covered enough ground already.